File: platform/polyglot/confluence_api/mjs/src/services/backup-service.mjs
/**
 * @module services/backup-service
 * @description Service for Confluence Backup & Restore REST API operations.
 *
 * Provides site and space backup/restore operations, job management (polling,
 * cancellation, download), and file-based restore uploads. Includes a polling
 * utility for monitoring long-running backup/restore jobs.
 *
 * Confluence Data Center REST API v9.2.3.
 */

import { createLogger } from '../logger.mjs';

const log = createLogger('confluence_api', import.meta.url);

/**
 * Service class for Confluence backup and restore operations.
 */
export class BackupService {
  /**
   * @param {import('../client/FetchClient.mjs').FetchClient} client
   *   The Confluence fetch client instance.
   */
  constructor(client) {
    /** @private */
    this._client = client;
  }

  // ---------------------------------------------------------------------------
  // Site Backup & Restore
  // ---------------------------------------------------------------------------

  /**
   * Initiate a full site backup.
   *
   * @param {Object} data - Backup configuration payload.
   * @param {boolean} [data.cbAttachments] - Whether to include attachments.
   * @returns {Promise<Object>} Backup job details.
   */
  async backupSite(data) {
    log.debug('backupSite called', { cbAttachments: data?.cbAttachments });
    try {
      const result = await this._client.post('backup/site', data);
      log.info('backupSite succeeded', { jobId: result.jobId ?? result.id });
      return result;
    } catch (error) {
      log.error('backupSite failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Initiate a site restore from a server-side backup file.
   *
   * @param {Object} data - Restore configuration payload.
   * @param {string} data.fileName - Name of the backup file on the server.
   * @returns {Promise<Object>} Restore job details.
   */
  async restoreSite(data) {
    log.debug('restoreSite called', { fileName: data?.fileName });
    try {
      const result = await this._client.post('backup/site/restore', data);
      log.info('restoreSite succeeded', { jobId: result.jobId ?? result.id });
      return result;
    } catch (error) {
      log.error('restoreSite failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Upload a backup file and initiate a site restore (multipart upload).
   *
   * @param {FormData} formData - Multipart form data containing the backup file.
   * @returns {Promise<Object>} Restore job details.
   */
  async restoreSiteUpload(formData) {
    log.debug('restoreSiteUpload called');
    try {
      const result = await this._client.post('backup/site/restore/upload', formData, {
        headers: { 'X-Atlassian-Token': 'nocheck' },
      });
      log.info('restoreSiteUpload succeeded', { jobId: result.jobId ?? result.id });
      return result;
    } catch (error) {
      log.error('restoreSiteUpload failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Space Backup & Restore
  // ---------------------------------------------------------------------------

  /**
   * Initiate a space backup.
   *
   * @param {Object} data - Backup configuration payload.
   * @param {string} data.spaceKey - The space key to back up.
   * @param {boolean} [data.cbAttachments] - Whether to include attachments.
   * @returns {Promise<Object>} Backup job details.
   */
  async backupSpace(data) {
    log.debug('backupSpace called', { spaceKey: data?.spaceKey });
    try {
      const result = await this._client.post('backup/space', data);
      log.info('backupSpace succeeded', { spaceKey: data?.spaceKey, jobId: result.jobId ?? result.id });
      return result;
    } catch (error) {
      log.error('backupSpace failed', { spaceKey: data?.spaceKey, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Initiate a space restore from a server-side backup file.
   *
   * @param {Object} data - Restore configuration payload.
   * @param {string} data.fileName - Name of the backup file on the server.
   * @returns {Promise<Object>} Restore job details.
   */
  async restoreSpace(data) {
    log.debug('restoreSpace called', { fileName: data?.fileName });
    try {
      const result = await this._client.post('backup/space/restore', data);
      log.info('restoreSpace succeeded', { jobId: result.jobId ?? result.id });
      return result;
    } catch (error) {
      log.error('restoreSpace failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Upload a backup file and initiate a space restore (multipart upload).
   *
   * @param {FormData} formData - Multipart form data containing the backup file.
   * @returns {Promise<Object>} Restore job details.
   */
  async restoreSpaceUpload(formData) {
    log.debug('restoreSpaceUpload called');
    try {
      const result = await this._client.post('backup/space/restore/upload', formData, {
        headers: { 'X-Atlassian-Token': 'nocheck' },
      });
      log.info('restoreSpaceUpload succeeded', { jobId: result.jobId ?? result.id });
      return result;
    } catch (error) {
      log.error('restoreSpaceUpload failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Job Management
  // ---------------------------------------------------------------------------

  /**
   * Get all backup/restore jobs.
   *
   * @returns {Promise<Object>} List of jobs.
   */
  async getJobs() {
    log.debug('getJobs called');
    try {
      const result = await this._client.get('backup/queue');
      log.info('getJobs succeeded', { count: result.results?.length ?? result.length });
      return result;
    } catch (error) {
      log.error('getJobs failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Get details of a specific backup/restore job.
   *
   * @param {string} jobId - The job ID.
   * @returns {Promise<Object>} Job details including status and progress.
   */
  async getJob(jobId) {
    log.debug('getJob called', { jobId });
    try {
      const result = await this._client.get(`backup/queue/${jobId}`);
      log.info('getJob succeeded', { jobId, status: result.status });
      return result;
    } catch (error) {
      log.error('getJob failed', { jobId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Download the output file of a completed backup job.
   *
   * Returns the raw Response object for binary streaming.
   *
   * @param {string} jobId - The job ID.
   * @returns {Promise<Response>} Raw HTTP response with the backup file body.
   */
  async downloadJob(jobId) {
    log.debug('downloadJob called', { jobId });
    try {
      const result = await this._client.getRaw(`backup/queue/${jobId}/download`);
      log.info('downloadJob succeeded', { jobId });
      return result;
    } catch (error) {
      log.error('downloadJob failed', { jobId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Clear all completed and failed jobs from the queue.
   *
   * @returns {Promise<void>}
   */
  async clearJobQueue() {
    log.debug('clearJobQueue called');
    try {
      const result = await this._client.delete('backup/queue');
      log.info('clearJobQueue succeeded');
      return result;
    } catch (error) {
      log.error('clearJobQueue failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * Cancel a running backup/restore job.
   *
   * @param {string} jobId - The job ID.
   * @returns {Promise<void>}
   */
  async cancelJob(jobId) {
    log.debug('cancelJob called', { jobId });
    try {
      const result = await this._client.post(`backup/queue/${jobId}/cancel`);
      log.info('cancelJob succeeded', { jobId });
      return result;
    } catch (error) {
      log.error('cancelJob failed', { jobId, message: error.message, status: error.status });
      throw error;
    }
  }

  /**
   * List available restore files on the server.
   *
   * @returns {Promise<Object>} List of restore file names.
   */
  async listRestoreFiles() {
    log.debug('listRestoreFiles called');
    try {
      const result = await this._client.get('backup/restore/files');
      log.info('listRestoreFiles succeeded');
      return result;
    } catch (error) {
      log.error('listRestoreFiles failed', { message: error.message, status: error.status });
      throw error;
    }
  }

  // ---------------------------------------------------------------------------
  // Polling Utility
  // ---------------------------------------------------------------------------

  /**
   * Poll a backup/restore job until it completes or times out.
   *
   * Periodically fetches job status and invokes an optional progress callback.
   * Resolves when the job reaches a terminal state (completed, failed, cancelled).
   * Rejects if the polling timeout is exceeded or the job fails.
   *
   * @param {string} jobId - The job ID to poll.
   * @param {Object} [options={}] - Polling options.
   * @param {number} [options.interval=2000] - Polling interval in milliseconds.
   * @param {number} [options.timeout=300000] - Maximum polling duration in milliseconds (default 5 minutes).
   * @param {Function} [options.onProgress] - Callback invoked with the job status on each poll.
   * @returns {Promise<Object>} The final job status object.
   * @throws {Error} If the job fails or the timeout is exceeded.
   */
  async pollJob(jobId, { interval = 2000, timeout = 300000, onProgress } = {}) {
    log.debug('pollJob called', { jobId, interval, timeout });
    const startTime = Date.now();

    while (true) {
      const job = await this.getJob(jobId);

      if (typeof onProgress === 'function') {
        onProgress(job);
      }

      const status = (job.status || '').toLowerCase();

      if (status === 'completed' || status === 'complete') {
        log.info('pollJob completed', { jobId, status: job.status });
        return job;
      }

      if (status === 'failed' || status === 'error') {
        const errorMsg = `Backup/restore job ${jobId} failed: ${job.message || job.status}`;
        log.error('pollJob job failed', { jobId, status: job.status, message: job.message });
        throw new Error(errorMsg);
      }

      if (status === 'cancelled' || status === 'canceled') {
        const cancelMsg = `Backup/restore job ${jobId} was cancelled`;
        log.info('pollJob job cancelled', { jobId });
        throw new Error(cancelMsg);
      }

      const elapsed = Date.now() - startTime;
      if (elapsed >= timeout) {
        const timeoutMsg = `Polling timed out after ${timeout}ms for job ${jobId}`;
        log.error('pollJob timed out', { jobId, elapsed, timeout });
        throw new Error(timeoutMsg);
      }

      await new Promise((resolve) => setTimeout(resolve, interval));
    }
  }
}

File: platform/polyglot/confluence_api/py/src/confluence_api/services/backup_service.py
"""
Backup and restore service for Confluence Data Center REST API v9.2.3.

Provides operations for site and space backup/restore, job management,
binary download of completed backups, and a polling utility for
long-running backup/restore jobs.
"""

from __future__ import annotations

import asyncio
import time
from collections.abc import Callable
from pathlib import Path
from typing import Any

from confluence_api.logger import create_logger
from confluence_api.multipart import build_multipart_files

log = create_logger("confluence_api", __file__)


class BackupService:
    """Service for Confluence backup and restore operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    # ── Site Backup / Restore ────────────────────────────────────────────

    def backup_site(self, data: dict) -> dict:
        """
        Initiate a full site backup.

        Args:
            data: Backup configuration (e.g. {"cbAttachments": True}).

        Returns:
            Job submission dict containing the job ID.
        """
        log.debug("backup_site called", {"data_keys": list(data.keys())})
        result = self._client.post("backup/site", json_data=data)
        log.info("backup_site succeeded", {"job_id": result.get("id")})
        return result

    def restore_site(self, data: dict) -> dict:
        """
        Initiate a site restore from a server-side file.

        Args:
            data: Restore configuration including the filename on the server.

        Returns:
            Job submission dict.
        """
        log.debug("restore_site called", {"data_keys": list(data.keys())})
        result = self._client.post("backup/site/restore", json_data=data)
        log.info("restore_site succeeded", {"job_id": result.get("id")})
        return result

    def restore_site_upload(self, file_path: str) -> dict:
        """
        Restore a site from an uploaded backup file (multipart upload).

        Args:
            file_path: Path to the backup file on disk.

        Returns:
            Job submission dict.
        """
        log.debug("restore_site_upload called", {"file_path": file_path})
        files = build_multipart_files(file_path)
        result = self._client.post("backup/site/restore/upload", files=files)
        log.info("restore_site_upload succeeded")
        return result

    # ── Space Backup / Restore ───────────────────────────────────────────

    def backup_space(self, data: dict) -> dict:
        """
        Initiate a space backup.

        Args:
            data: Backup configuration including space key.

        Returns:
            Job submission dict.
        """
        log.debug("backup_space called", {"space_key": data.get("spaceKey")})
        result = self._client.post("backup/space", json_data=data)
        log.info("backup_space succeeded", {"job_id": result.get("id")})
        return result

    def restore_space(self, data: dict) -> dict:
        """
        Initiate a space restore from a server-side file.

        Args:
            data: Restore configuration including the filename on the server.

        Returns:
            Job submission dict.
        """
        log.debug("restore_space called", {"data_keys": list(data.keys())})
        result = self._client.post("backup/space/restore", json_data=data)
        log.info("restore_space succeeded", {"job_id": result.get("id")})
        return result

    def restore_space_upload(self, file_path: str) -> dict:
        """
        Restore a space from an uploaded backup file (multipart upload).

        Args:
            file_path: Path to the backup file on disk.

        Returns:
            Job submission dict.
        """
        log.debug("restore_space_upload called", {"file_path": file_path})
        files = build_multipart_files(file_path)
        result = self._client.post("backup/space/restore/upload", files=files)
        log.info("restore_space_upload succeeded")
        return result

    # ── Job Management ───────────────────────────────────────────────────

    def get_jobs(self) -> dict:
        """Retrieve all backup/restore jobs."""
        log.debug("get_jobs called")
        result = self._client.get("backup/jobs")
        log.info("get_jobs succeeded")
        return result

    def get_job(self, job_id: str) -> dict:
        """Retrieve the status of a specific backup/restore job."""
        log.debug("get_job called", {"job_id": job_id})
        result = self._client.get(f"backup/jobs/{job_id}")
        log.info("get_job succeeded", {"job_id": job_id})
        return result

    def download_job(self, job_id: str) -> bytes:
        """
        Download the completed backup file for a job.

        Returns raw bytes of the backup file. Uses the client's get_raw
        method to avoid JSON parsing of binary data.

        Args:
            job_id: The job ID whose output file to download.

        Returns:
            Raw bytes of the backup archive.
        """
        log.debug("download_job called", {"job_id": job_id})
        response = self._client.get_raw(f"backup/jobs/{job_id}/download")
        content: bytes = response.content
        log.info("download_job succeeded", {
            "job_id": job_id,
            "size_bytes": len(content),
        })
        return content

    def clear_job_queue(self) -> dict:
        """Clear all completed jobs from the queue."""
        log.debug("clear_job_queue called")
        result = self._client.delete("backup/jobs")
        log.info("clear_job_queue succeeded")
        return result

    def cancel_job(self, job_id: str) -> dict:
        """Cancel a running backup/restore job."""
        log.debug("cancel_job called", {"job_id": job_id})
        result = self._client.delete(f"backup/jobs/{job_id}")
        log.info("cancel_job succeeded", {"job_id": job_id})
        return result

    def list_restore_files(self) -> dict:
        """List available restore files on the server."""
        log.debug("list_restore_files called")
        result = self._client.get("backup/restore/files")
        log.info("list_restore_files succeeded")
        return result

    # ── Polling Utility ──────────────────────────────────────────────────

    def poll_job(
        self,
        job_id: str,
        interval: float = 2.0,
        timeout: float = 300.0,
        on_progress: Callable[[dict], None] | None = None,
    ) -> dict:
        """
        Poll a backup/restore job until completion or timeout.

        Periodically checks the job status and optionally invokes a progress
        callback. Returns the final job status dict.

        Args:
            job_id: The job ID to poll.
            interval: Seconds between poll requests (default 2.0).
            timeout: Maximum seconds to wait before raising TimeoutError (default 300.0).
            on_progress: Optional callback invoked with the job status dict on each poll.

        Returns:
            Final job status dict.

        Raises:
            TimeoutError: If the job does not complete within the timeout period.
        """
        log.debug("poll_job started", {
            "job_id": job_id,
            "interval": interval,
            "timeout": timeout,
        })
        start_time = time.monotonic()

        while True:
            elapsed = time.monotonic() - start_time
            if elapsed >= timeout:
                log.error("poll_job timed out", {"job_id": job_id, "elapsed": elapsed})
                raise TimeoutError(
                    f"Backup/restore job '{job_id}' did not complete within {timeout}s"
                )

            job_status = self.get_job(job_id)

            if on_progress is not None:
                on_progress(job_status)

            state = job_status.get("state", "").upper()

            if state in ("COMPLETE", "COMPLETED", "SUCCESS"):
                log.info("poll_job completed successfully", {
                    "job_id": job_id,
                    "elapsed": round(elapsed, 2),
                })
                return job_status

            if state in ("FAILED", "ERROR", "CANCELLED"):
                log.error("poll_job ended with failure state", {
                    "job_id": job_id,
                    "state": state,
                    "elapsed": round(elapsed, 2),
                })
                return job_status

            log.debug("poll_job waiting", {
                "job_id": job_id,
                "state": state,
                "elapsed": round(elapsed, 2),
            })
            time.sleep(interval)


class AsyncBackupService:
    """Async service for Confluence backup and restore operations."""

    def __init__(self, client: Any) -> None:
        self._client = client

    # ── Site Backup / Restore ────────────────────────────────────────────

    async def backup_site(self, data: dict) -> dict:
        """Initiate a full site backup."""
        log.debug("backup_site called", {"data_keys": list(data.keys())})
        result = await self._client.post("backup/site", json_data=data)
        log.info("backup_site succeeded", {"job_id": result.get("id")})
        return result

    async def restore_site(self, data: dict) -> dict:
        """Initiate a site restore from a server-side file."""
        log.debug("restore_site called", {"data_keys": list(data.keys())})
        result = await self._client.post("backup/site/restore", json_data=data)
        log.info("restore_site succeeded", {"job_id": result.get("id")})
        return result

    async def restore_site_upload(self, file_path: str) -> dict:
        """Restore a site from an uploaded backup file (multipart upload)."""
        log.debug("restore_site_upload called", {"file_path": file_path})
        files = build_multipart_files(file_path)
        result = await self._client.post("backup/site/restore/upload", files=files)
        log.info("restore_site_upload succeeded")
        return result

    # ── Space Backup / Restore ───────────────────────────────────────────

    async def backup_space(self, data: dict) -> dict:
        """Initiate a space backup."""
        log.debug("backup_space called", {"space_key": data.get("spaceKey")})
        result = await self._client.post("backup/space", json_data=data)
        log.info("backup_space succeeded", {"job_id": result.get("id")})
        return result

    async def restore_space(self, data: dict) -> dict:
        """Initiate a space restore from a server-side file."""
        log.debug("restore_space called", {"data_keys": list(data.keys())})
        result = await self._client.post("backup/space/restore", json_data=data)
        log.info("restore_space succeeded", {"job_id": result.get("id")})
        return result

    async def restore_space_upload(self, file_path: str) -> dict:
        """Restore a space from an uploaded backup file (multipart upload)."""
        log.debug("restore_space_upload called", {"file_path": file_path})
        files = build_multipart_files(file_path)
        result = await self._client.post("backup/space/restore/upload", files=files)
        log.info("restore_space_upload succeeded")
        return result

    # ── Job Management ───────────────────────────────────────────────────

    async def get_jobs(self) -> dict:
        """Retrieve all backup/restore jobs."""
        log.debug("get_jobs called")
        result = await self._client.get("backup/jobs")
        log.info("get_jobs succeeded")
        return result

    async def get_job(self, job_id: str) -> dict:
        """Retrieve the status of a specific backup/restore job."""
        log.debug("get_job called", {"job_id": job_id})
        result = await self._client.get(f"backup/jobs/{job_id}")
        log.info("get_job succeeded", {"job_id": job_id})
        return result

    async def download_job(self, job_id: str) -> bytes:
        """
        Download the completed backup file for a job.

        Returns raw bytes of the backup file. Uses the client's get_raw
        method to avoid JSON parsing of binary data.

        Args:
            job_id: The job ID whose output file to download.

        Returns:
            Raw bytes of the backup archive.
        """
        log.debug("download_job called", {"job_id": job_id})
        response = await self._client.get_raw(f"backup/jobs/{job_id}/download")
        content: bytes = response.content
        log.info("download_job succeeded", {
            "job_id": job_id,
            "size_bytes": len(content),
        })
        return content

    async def clear_job_queue(self) -> dict:
        """Clear all completed jobs from the queue."""
        log.debug("clear_job_queue called")
        result = await self._client.delete("backup/jobs")
        log.info("clear_job_queue succeeded")
        return result

    async def cancel_job(self, job_id: str) -> dict:
        """Cancel a running backup/restore job."""
        log.debug("cancel_job called", {"job_id": job_id})
        result = await self._client.delete(f"backup/jobs/{job_id}")
        log.info("cancel_job succeeded", {"job_id": job_id})
        return result

    async def list_restore_files(self) -> dict:
        """List available restore files on the server."""
        log.debug("list_restore_files called")
        result = await self._client.get("backup/restore/files")
        log.info("list_restore_files succeeded")
        return result

    # ── Polling Utility ──────────────────────────────────────────────────

    async def poll_job(
        self,
        job_id: str,
        interval: float = 2.0,
        timeout: float = 300.0,
        on_progress: Callable[[dict], None] | None = None,
    ) -> dict:
        """
        Poll a backup/restore job until completion or timeout.

        Periodically checks the job status and optionally invokes a progress
        callback. Returns the final job status dict.

        Args:
            job_id: The job ID to poll.
            interval: Seconds between poll requests (default 2.0).
            timeout: Maximum seconds to wait before raising TimeoutError (default 300.0).
            on_progress: Optional callback invoked with the job status dict on each poll.

        Returns:
            Final job status dict.

        Raises:
            TimeoutError: If the job does not complete within the timeout period.
        """
        log.debug("poll_job started", {
            "job_id": job_id,
            "interval": interval,
            "timeout": timeout,
        })
        start_time = time.monotonic()

        while True:
            elapsed = time.monotonic() - start_time
            if elapsed >= timeout:
                log.error("poll_job timed out", {"job_id": job_id, "elapsed": elapsed})
                raise TimeoutError(
                    f"Backup/restore job '{job_id}' did not complete within {timeout}s"
                )

            job_status = await self.get_job(job_id)

            if on_progress is not None:
                on_progress(job_status)

            state = job_status.get("state", "").upper()

            if state in ("COMPLETE", "COMPLETED", "SUCCESS"):
                log.info("poll_job completed successfully", {
                    "job_id": job_id,
                    "elapsed": round(elapsed, 2),
                })
                return job_status

            if state in ("FAILED", "ERROR", "CANCELLED"):
                log.error("poll_job ended with failure state", {
                    "job_id": job_id,
                    "state": state,
                    "elapsed": round(elapsed, 2),
                })
                return job_status

            log.debug("poll_job waiting", {
                "job_id": job_id,
                "state": state,
                "elapsed": round(elapsed, 2),
            })
            await asyncio.sleep(interval)
